### ISSUE TITLE
Parallel="classes" executes methods of test class in different threads

### ISSUE THREAD

**Issue by ryudice** [cbeust/testng#1773]
### TestNG Version
6.14.3

### Expected behavior
When using parallel="classes" attribute in a suite XML file and using either the "dependsOnMethods" or the "priority" attribute on the test annotation it should run classes in parallel and the methods of each of the classes in the same thread

### Actual behavior
Using parallel="classes" on test methods with the priority or dependsOnMethods properties defined causes each of the methods of the test class to run in different thread, eventhough they seem to run sequentially.

### Is the issue reproductible on runner?
I'm passing the suite xml file to the testng runner.

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
If you create a test class like this: 

```java
public class PriorityTest2 {

    @Test(priority = 1)
    public void FirstTest(){
        System.out.println("First test ran");
        System.out.println(Thread.currentThread().getId());
    }

    @Test(priority = 2)
    public void SecondTest(){
        System.out.println("Second test ran");
        System.out.println(Thread.currentThread().getId());
    }

    @Test(priority = 3)
    public void ThridTest(){
        System.out.println("Third test ran");
        System.out.println(Thread.currentThread().getId());
    }
}
```

and then a suite XML file like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<suite name="Mozilla Test Suite" parallel="classes" thread-count="6" >
    <test name="Test 1" group-by-instances="true" >
        <classes>
            <class name = "PriorityTest"/>
        </classes>
    </test>
</suite>
```

You get the following output:
```
First test ran
12
Second test ran
13
Third test ran
14
```

However if you change the test suite to use parallel='tests', then you get the correct output, which is each test method running in the same thread:

```
First test ran
12
Second test ran
12
Third test ran
12
```

**Comment by tetienne** [comment]
Seems a duplicate of #1185

**Comment by juherr** [comment]
@tetienne I'm not sure it is exactly the same issue. Here priority is used which is a global suite dependency feature and we have a strange behavior with parallel=test.
But I agree, it's probably the same root cause :)

**Comment by krmahadevan** [comment]
@juherr - I spent some time looking at this.The root cause is the same. 
When TestNG encounters methods with dependencies, we end up running independent methods separately and then followed by dependent methods. When parallelism is involved in this combo, we end up making use of the `org.testng.internal.thread.graph.GraphThreadPoolExecutor` implementation for running the tasks. Since `ThreadPoolExecutor` variants doesn't let us define thread affinity to our tasks, we are kind of stuck with a situation wherein dependent methods end up getting lined up on a different thread than their parent tasks, resulting in this issue.

The easiest way for us to solve this, without having to change `DependencyGraph` implementation would be to find a way to define thread affinity (i.e., run a dependent task only in the same thread as the one in which the independent one ran). I can think of some hacks in mind viz.,

1. Start with executing `org.testng.internal.thread.graph.GraphThreadPoolExecutor#runNodes` in its current fashion.
2. Enrich `org.testng.internal.thread.graph.IWorker` to retrieve Thread ID on which it was executed.
3. Enrich `org.testng.internal.thread.graph.IWorker` to retrieve a status which represents if the task was run to completion.
4. Enrich `org.testng.internal.thread.graph.IWorker` implementations viz., `org.testng.internal.TestMethodWorker` and `org.testng.SuiteRunnerWorker` to be able to be injected with the Thread ID of the independent task, and use that as a cross comparison with the `Thread.currentThread().getId()` to see if there's a match and only then execute its `run()` method.
5. Now within the `org.testng.internal.thread.graph.GraphThreadPoolExecutor#afterExecute` we query the `IWorker` to find out if it ran to completion via (3) and only then set the status.

I am still contemplating on whether this entire thing would look like an ugly hack :)

Would be great if you could at least let me know if this overall approach sounds fine and can even be attempted or not. Dont want to spend time on something which doesn't even make sense.

@cbeust - It would be great if you could please add your thoughts as well.

**Comment by juherr** [comment]
@krmahadevan I think the best way is trying it and it will easier to discuss with some code ;)
Another idea could be to add an extra `level` parameter to `getFreeNodes()` in order to skip edges with a lower weight. But I didn't test it and I don't know if it can work.

**Comment by krmahadevan** [comment]
Fixed by #1783
